components/gemini/browser/gemini_service.cc

```diff
@@ -0,0 +1,488 @@
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/utf_string_conversions.h"
+#include "base/task/post_task.h"
+#include "base/task_runner_util.h"
+#include "base/time/time.h"
+#include "base/token.h"
+#include "brave/common/pref_names.h"
+#include "components/os_crypt/os_crypt.h"
+#include "components/prefs/pref_service.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+#include "net/base/load_flags.h"
+#include "net/base/url_util.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "brave/components/gemini/browser/gemini_json_parser.h"
+#include "base/json/json_writer.h"
+#include "crypto/random.h"
+
+namespace {
+  const char oauth_host[] = "exchange.gemini.com";
+  const char api_host[] = "api.gemini.com";
+  const char oauth_callback[] = "com.brave.gemini://authorization";
+  const char oauth_scope[] = "trader";
+  const char oauth_url[] = "https://exchange.gemini.com/auth";
+  const unsigned int kRetriesCountOnNetworkChange = 1;
+
+  net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+    return net::DefineNetworkTrafficAnnotation("gemini_service", R"(
+        semantics {
+          sender: "Gemini Service"
+          description:
+            "This service is used to communicate with Gemini "
+            "on behalf of the user interacting with the Gemini widget."
+          trigger:
+            "Triggered by user connecting the Gemini widget."
+          data:
+            "Account information, balances"
+          destination: WEBSITE
+        }
+        policy {
+          cookies_allowed: NO
+          setting:
+            "You can enable or disable this feature on the new tab page."
+          policy_exception_justification:
+            "Not implemented."
+        }
+      )");
+  }
+
+  GURL GetURLWithPath(const std::string& host, const std::string& path) {
+    return GURL(std::string(url::kHttpsScheme) + "://" + host).Resolve(path);
+  }
+
+  std::string CreateJSONRequestBody(const base::Value& dict) {
+    std::string json;
+    base::JSONWriter::Write(dict, &json);
+    return json;
+  }
+
+  std::string GetEncodedRequestPayload(std::string payload) {
+    std::string encoded_payload;
+    std::string request_payload =
+      "{\"request\": \"" + payload + "\"}";
+    base::Base64Encode(request_payload, &encoded_payload);
+    return encoded_payload;
+  }
+
+  std::string GetEncodedExecutePayload(std::string symbol,
+                                       std::string side,
+                                       std::string quantity,
+                                       std::string price,
+                                       std::string fee,
+                                       int quote_id) {
+    std::string encoded_payload;
+    std::string request_payload = "{\"request\": \"" +
```

Comment:
same as https://github.com/brave/brave-core/pull/6062/files#r454587791 above

Reply:
now using a dictionary and JSONWriter